lib/matplotlib/collections.py

```diff
@@ -422,24 +420,71 @@ def draw(self, renderer):
+            # Since third-party implementations of draw_path_collection() may not be
+            # introspectable, e.g. with inspect.signature, the only way is to try and
+            # call this with the hatchcolors parameter.
+            hatchcolors_arg_supported = True
+            try:
                 renderer.draw_path_collection(
-                    gc, transform.frozen(), ipaths,
+                    gc, transform.frozen(), [],
                     self.get_transforms(), offsets, offset_trf,
-                    [mcolors.to_rgba("none")], self._gapcolor,
-                    self._linewidths, ilinestyles,
+                    self.get_facecolor(), self.get_edgecolor(),
+                    self._linewidths, self._linestyles,
                     self._antialiaseds, self._urls,
-                    "screen")
+                    "screen", hatchcolors=self.get_hatchcolor()
+                )
+            except TypeError:
+                # If the renderer does not support the hatchcolors argument,
+                # it will raise a TypeError. In this case, we will
+                # iterate over all paths and draw them one by one.
+                hatchcolors_arg_supported = False
 
-            renderer.draw_path_collection(
-                gc, transform.frozen(), paths,
-                self.get_transforms(), offsets, offset_trf,
-                self.get_facecolor(), self.get_edgecolor(),
-                self._linewidths, self._linestyles,
-                self._antialiaseds, self._urls,
-                "screen")  # offset_position, kept for backcompat.
+            if self._gapcolor is not None:
+                # First draw paths within the gaps.
+                ipaths, ilinestyles = self._get_inverse_paths_linestyles()
+                args = [offsets, offset_trf, [mcolors.to_rgba("none")], self._gapcolor,
+                        self._linewidths, ilinestyles, self._antialiaseds, self._urls,
+                        "screen"]
+
+                if hatchcolors_arg_supported:
+                    renderer.draw_path_collection(gc, transform.frozen(), ipaths,
+                                                  self.get_transforms(), *args,
+                                                  hatchcolors=self.get_hatchcolor())
+                else:
+                    path_ids = renderer._iter_collection_raw_paths(
+                        transform.frozen(), ipaths, self.get_transforms())
+                    for xo, yo, path_id, gc0, rgbFace in renderer._iter_collection(
+                        gc, list(path_ids), *args, hatchcolors=self.get_hatchcolor(),
+                    ):
+                        path, transform = path_id
+                        if xo != 0 or yo != 0:
+                            transform = transform.frozen()
+                            transform.translate(xo, yo)
+                        renderer.draw_path(gc0, path, transform, rgbFace)
```

Comment:
For better backward-compatibility/performance, would it be reasonable to split the "backend does not support hatchcolors" into
```
if no_hatches_needed:
    renderer_draw_path_collection(...) # no hatch colors passed
else:
    # unroll the collection to draw paths in a loop
```
I'm unclear how performance-critical this is and whether we can unversally force the collection unrolling on backends that do not support hatchcolors.

ping @anntzer

Reply:
Good catch.  I agree that only going through the unrolled version when hatchcolors are actually passed would be much better.